usersession/userd/privileged_desktop_launcher.go

```diff
@@ -0,0 +1,307 @@
+// the XDG Base Directory specification.
+func desktopFileSearchPath() []string {
+	var desktopDirs []string
+
+	// First check $XDG_DATA_HOME, which defaults to $HOME/.local/share
+	dataHome := os.Getenv("XDG_DATA_HOME")
+	if dataHome == "" {
+		homeDir := os.Getenv("HOME")
+		if homeDir != "" {
+			dataHome = filepath.Join(homeDir, ".local/share")
+		}
+	}
+	if dataHome != "" {
+		desktopDirs = append(desktopDirs, filepath.Join(dataHome, "applications"))
+	}
+
+	// Next check $XDG_DATA_DIRS, with default from spec
+	dataDirs := os.Getenv("XDG_DATA_DIRS")
+	if dataDirs == "" {
+		dataDirs = "/usr/local/share/:/usr/share/"
+	}
+	for _, dir := range strings.Split(dataDirs, ":") {
+		if dir == "" {
+			continue
+		}
+		desktopDirs = append(desktopDirs, filepath.Join(dir, "applications"))
+	}
+
+	return desktopDirs
+}
+
+// findDesktopFile recursively tries each subdirectory that can be formed from the (split) desktop file ID.
+// Per https://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#desktop-file-id,
+// if desktop entries have dashes in the name ('-'), this could be an indication of subdirectories, so search
+// for those too. Eg, given foo-bar_baz_norf.desktop the following are searched for:
+//   o .../foo-bar_baz-norf.desktop
+//   o .../foo/bar_baz-norf.desktop
+//   o .../foo/bar_baz/norf.desktop
+//   o .../foo-bar_baz/norf.desktop
+// We're not required to diagnose multiple files matching the desktop file ID.
+func findDesktopFile(baseDir string, splitFileId []string) (string, error) {
+	desktopFile := filepath.Join(baseDir, strings.Join(splitFileId, "-"))
+
+	exists, isReg, _ := regularFileExists(desktopFile)
+	if exists && isReg {
+		return desktopFile, nil
+	}
+
+	// Iterate through the potential subdirectories formed by the first i elements of the desktop file ID.
+	for i := 1; i != len(splitFileId); i++ {
+		prefix := strings.Join(splitFileId[:i], "-")
+		// Don't treat empty or "." components as directory
+		// prefixes.  The ".." case is already filtered out by
+		// the isValidDesktopFileID regexp.
+		if prefix == "" || prefix == "." {
+			continue
+		}
+		desktopFile, err := findDesktopFile(filepath.Join(baseDir, prefix), splitFileId[i:])
+		if err == nil {
+			return desktopFile, nil
+		}
+	}
+
+	return "", fmt.Errorf("could not find desktop file")
+}
+
+// isValidDesktopFileID is based on the "File naming" section of the
+// Desktop Entry Specification, without the restriction on components
+// not starting with a digit (which desktop files created by snapd may
+// not satisfy).
+var isValidDesktopFileID = regexp.MustCompile(`^[A-Za-z0-9-_]+(\.[A-Za-z0-9-_]+)*.desktop$`).MatchString
+
+// desktopFileIDToFilename determines the path associated with a desktop file ID.
+func desktopFileIDToFilename(desktopFileID string) (string, error) {
+	if !isValidDesktopFileID(desktopFileID) {
+		return "", fmt.Errorf("cannot find desktop file for %q", desktopFileID)
```

Comment:
should this error message indicate that the desktop file ID is invalid insofar as it doesn't match the expected regex? something like

```suggestion
		return "", fmt.Errorf("cannot find desktop file for %q: desktop file ID is malformed", desktopFileID)
```

also fine for followup if we want it